File: frontend/src/api/slot.ts
import { apiClient } from './apiClient';

export interface Slot {
  id: string;
  office_id: string;
  slot_date: string;
  slot_time: string;
  capacity: number;
  taken: number;
  available: number;
}

export interface SlotSearchParams {
  location?: string;
  date?: string;
  page?: number;
  size?: number;
}

export interface PaginatedSlotResponse {
  items: Slot[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export const slotApi = {
  /**
   * Search for available slots by location and date
   * @param params Search parameters (location, date, page, size)
   * @returns Promise with paginated slot search results
   */
  searchSlots: async (params: SlotSearchParams): Promise<PaginatedSlotResponse> => {
    const queryParams = new URLSearchParams();

    if (params.location) queryParams.append('location', params.location);
    if (params.date) queryParams.append('date', params.date);
    if (params.page) queryParams.append('page', params.page.toString());
    if (params.size) queryParams.append('size', params.size.toString());

    const response = await apiClient.get(`/api/slots?${queryParams.toString()}`);
    return response.data;
  }
};

File: backend/app/schemas/slot.py
from typing import List, Optional
from pydantic import BaseModel
from datetime import date, time
from uuid import UUID

class SlotBase(BaseModel):
    office_id: UUID
    slot_date: date
    slot_time: time
    capacity: int

class SlotCreate(SlotBase):
    pass

class SlotResponse(SlotBase):
    id: UUID
    taken: int
    available: int

    class Config:
        orm_mode = True

class PaginatedSlotResponse(BaseModel):
    items: List[SlotResponse]
    total: int
    page: int
    size: int
    pages: int
